godot-core/src/obj/raw.rs

```diff
@@ -0,0 +1,605 @@
+    pub(super) fn instance_id_unchecked(&self) -> Option<InstanceId> {
+        self.cached_instance_id
+    }
+
+    /// Returns `true` if this object is neither dead nor null.
+    pub(super) fn is_instance_valid(&self) -> bool {
+        self.instance_id_or_none().is_some()
+    }
+
+    // See use-site for explanation.
+    fn is_cast_valid<U>(&self) -> bool
+    where
+        U: GodotClass,
+    {
+        if self.is_null() {
+            // Null can be cast to anything.
+            return true;
+        }
+
+        let as_obj =
+            unsafe { self.ffi_cast::<engine::Object>() }.expect("Everything inherits object");
+        let cast_is_valid = as_obj
+            .as_target()
+            .is_class(U::class_name().to_godot_string());
+        std::mem::forget(as_obj);
+        cast_is_valid
+    }
+
+    /// Returns `Ok(cast_obj)` on success, `Err(self)` on error
+    pub(super) fn owned_cast<U>(self) -> Result<RawGd<U>, Self>
+    where
+        U: GodotClass,
+    {
+        // Workaround for bug in Godot 4.0 that makes casts always succeed (https://github.com/godot-rust/gdext/issues/158).
+        // TODO once fixed in Godot, use #[cfg(before_api = "4.1")]
+        if !self.is_cast_valid::<U>() {
+            return Err(self);
+        }
+
+        // The unsafe { std::mem::transmute<&T, &Base>(self.inner()) } relies on the C++ static_cast class casts
+        // to return the same pointer, however in theory those may yield a different pointer (VTable offset,
+        // virtual inheritance etc.). It *seems* to work so far, but this is no indication it's not UB.
+        //
+        // The Deref/DerefMut impls for T implement an "implicit upcast" on the object (not Gd) level and
+        // rely on this (e.g. &Node3D -> &Node).
+
+        let result = unsafe { self.ffi_cast::<U>() };
+        match result {
+            Some(cast_obj) => {
+                // duplicated ref, one must be wiped
+                std::mem::forget(self);
+                Ok(cast_obj)
+            }
+            None => Err(self),
+        }
+    }
+
+    /// # Safety
+    /// Does not transfer ownership and is thus unsafe. Also operates on shared ref. Either the parameter or
+    /// the return value *must* be forgotten (since reference counts are not updated).
+    pub(super) unsafe fn ffi_cast<U>(&self) -> Option<RawGd<U>>
+    where
+        U: GodotClass,
+    {
+        if self.is_null() {
+            // Null can be cast to anything.
+            // Forgetting a null doesn't do anything, since dropping a null also does nothing.
+            return Some(RawGd::new_null());
+        }
+
+        let class_tag = interface_fn!(classdb_get_class_tag)(U::class_name().string_sys());
+        let cast_object_ptr = interface_fn!(object_cast_to)(self.obj_sys(), class_tag);
+
+        // Create weak object, as ownership will be moved and reference-counter stays the same
+        sys::ptr_then(cast_object_ptr, |ptr| RawGd::from_obj_sys_weak(ptr))
+    }
```

Comment:
Basically unchanged except null-check
```rs
 unsafe fn ffi_cast<U>(&self) -> Option<Gd<U>>
    where
        U: GodotClass,
    {
        let class_tag = interface_fn!(classdb_get_class_tag)(U::class_name().string_sys());
        let cast_object_ptr = interface_fn!(object_cast_to)(self.obj_sys(), class_tag);

        // Create weak object, as ownership will be moved and reference-counter stays the same
        sys::ptr_then(cast_object_ptr, |ptr| Gd::from_obj_sys_weak(ptr))
    }
```

Reply:
I'm curious, when do we need to cast between null objects?

A `Gd<T>` can't hold nulls, and `Option<Gd<T>>` would be `None`.

Reply:
Similar thoughts apply to other methods doing null checks. If we don't expect those, an `assert!(!null)` could potentially prevent logic errors.

Reply:
We use casting when we turn `Variant` into `RawGd`, and so if you turn a `Variant` into an `Option<Gd<T>>` then you'd first turn the `Variant` into a `RawGd<Object>`, then cast that into `RawGd<T>` then turn that into `Option<Gd<T>>`. So if the `Variant` is null we'd be casting a null object.

Reply:
Makes sense! Can you write down this conversion chain in a comment?  
It's enough to do it here in `ffi_cast`.